contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1057 @@
+        }
+    }
+
+    /**
+     * Changes the principal amount. This changes both the variable to track the total
+     * amount as well as the variable to track a particular bucket.
+     *
+     * @param  bucket    The bucket number
+     * @param  amount    The amount to change the principal amount by
+     * @param  increase  True if positive change, false if negative change
+     */
+    function changePrincipal(
+        uint256 bucket,
+        uint256 amount,
+        bool increase
+    )
+        internal
+    {
+        if (amount == 0) {
+            return;
+        }
+
+        if (increase) {
+            principalTotal = principalTotal.add(amount);
+            principalForBucket[bucket] = principalForBucket[bucket].add(amount);
+        } else {
+            principalTotal = principalTotal.sub(amount);
+            principalForBucket[bucket] = principalForBucket[bucket].sub(amount);
+        }
+    }
+
+    /**
+     * Increases the 'weight' values for a bucket and an account within that bucket
+     *
+     * @param  bucket       The bucket number
+     * @param  account      The account to remove weight from
+     * @param  weightToAdd  Adds this amount of weight
+     */
+    function accountForDeposit(
+        uint256 bucket,
+        address account,
+        uint256 weightToAdd
+    )
+        internal
+    {
+        weightForBucketForAccount[bucket][account] =
+            weightForBucketForAccount[bucket][account].add(weightToAdd);
+        weightForBucket[bucket] = weightForBucket[bucket].add(weightToAdd);
+    }
+
+    /**
+     * Decreases the 'weight' values for a bucket and an account within that bucket.
+     *
+     * @param  bucket         The bucket number
+     * @param  account        The account to remove weight from
+     * @param  maximumWeight  Removes up-to this amount of weight
+     * @return                The amount of weight removed
+     */
+    function accountForWithdraw(
+        uint256 bucket,
+        address account,
+        uint256 maximumWeight
+    )
+        internal
+        returns (uint256)
+    {
+        uint256 userWeight = weightForBucketForAccount[bucket][account];
+        uint256 weightToWithdraw = Math.min256(userWeight, maximumWeight);
+
+        weightForBucket[bucket] = weightForBucket[bucket].sub(weightToWithdraw);
+        weightForBucketForAccount[bucket][account] = userWeight.sub(weightToWithdraw);
+
+        return weightToWithdraw;
+    }
+
+    // ============ Getter Functions ============
+
+    /**
+     * Get the current bucket number that funds will be deposited into. This is the highest bucket
+     * so far. All lent funds before the position open will go into bucket 0. All lent funds after
+     * position open will go into buckets 1+.
+     */
+    function getBucketNumber()
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 marginTimestamp = Margin(DYDX_MARGIN).getPositionStartTimestamp(POSITION_ID);
```

Comment:
maybe assert the position hasn't been closed? It would be bad if it was closed and we started overwriting the stuff in bucket 0

Reply:
done